chia/full_node/full_node.py

```diff
@@ -2351,26 +2343,71 @@ async def add_transaction(
                 # vector.
                 mempool_item = self.mempool_manager.get_mempool_item(spend_name)
                 assert mempool_item is not None
-                fees = mempool_item.fee
-                assert fees >= 0
-                assert cost is not None
-                new_tx = full_node_protocol.NewTransaction(
-                    spend_name,
-                    cost,
-                    fees,
-                )
-                msg = make_msg(ProtocolMessageTypes.new_transaction, new_tx)
-                if peer is None:
-                    await self.server.send_to_all([msg], NodeType.FULL_NODE)
-                else:
-                    await self.server.send_to_all([msg], NodeType.FULL_NODE, peer.peer_node_id)
+
+                await self.notify_peers_of_mempool_item(mempool_item, peer)
+
                 if self.simulator_transaction_callback is not None:  # callback
                     await self.simulator_transaction_callback(spend_name)  # pylint: disable=E1102
+
             else:
                 self.mempool_manager.remove_seen(spend_name)
                 self.log.debug(f"Wasn't able to add transaction with id {spend_name}, status {status} error: {error}")
         return status, error
 
+    async def notify_peers_of_mempool_item(
+        self, mempool_item: MempoolItem, peer: Optional[WSChiaConnection] = None
+    ) -> None:
+        new_tx = full_node_protocol.NewTransaction(
+            mempool_item.spend_bundle_name,
+            mempool_item.cost,
+            mempool_item.fee,
+        )
+        new_tx_msg = make_msg(ProtocolMessageTypes.new_transaction, new_tx)
+
+        await self.server.send_to_all([new_tx_msg], NodeType.FULL_NODE, None if peer is None else peer.peer_node_id)
+
+        wallet_update = wallet_protocol.TransactionAddedUpdate(
+            mempool_item.spend_bundle_name,
+            mempool_item.height_added_to_mempool,
+            mempool_item.cost,
+            mempool_item.additions,
+            mempool_item.removals,
+        )
+        wallet_update_msg = make_msg(ProtocolMessageTypes.transaction_added_update, wallet_update)
+
+        peers = await self.peers_for_mempool_item(mempool_item)
+
+        def new_clients(conn: WSChiaConnection) -> bool:
+            return conn.peer_node_id in peers and conn.protocol_version > Version("0.0.36")
+
+        await self.server.send_to_all_if(
+            [wallet_update_msg], NodeType.WALLET, new_clients, None if peer is None else peer.peer_node_id
```

Comment:
it seems questionable that we send this new message to you unsolicited. At least you've only subscribed to blockchain updates, not explicitly to mempool updates. Although, arguably you would be interested in these updates too, most likely.

Did you consider a protocol where the wallet explicitly opts-in to receiving these updates?
If you do that, you also don't need to filter based on protocol version, only wallets supporting the new protocol will have opt-in.

Also, @almogdepaz prefers advertizing support for a new protocol using capabilities. I think this is a case where that might make sense, since it's not replacing an old protocol (that we would like to get rid of in the future).

Reply:
yeah, not advertising the capability would be opting out

Reply:
This will be moving to another PR later.